### PR TITLE
refactor: migrate SqlEditor component to FC & tsx

### PR DESCRIPTION
**Pull request by EugeneTorap**
### SUMMARY
Migrated SqlEditor to TypeScript to apply direction outlined in #18100 .

### ADDITIONAL INFORMATION

- [x] Has associated issue: #18100
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by EugeneTorap** [comment]
@ad-m @eschutho I have a issue with `<Menu onClick={this.handleMenuClick} style={{ width: 176 }}>`
SqlEditor class component doesn't have `handleMenuClick`. What should I do?

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+import Mousetrap from 'mousetrap';
+import Button from 'src/components/Button';
+import Timer from 'src/components/Timer';
+import {
+  Dropdown,
+  Menu as AntdMenu,
+  Menu,
+  Switch,
+  Input,
+} from 'src/common/components';
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:36]: The components `Dropdown` and `Switch` are available in the `src/components` directory. I think we should prefer those to keep consistency.

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+ * under the License.
+ */
+import React, { FC, useState, useEffect, useRef } from 'react';
+import { CSSTransition } from 'react-transition-group';
+import { useDispatch, useSelector } from 'react-redux';
+import Split from 'react-split';
+import { t, styled, useTheme } from '@superset-ui/core';
+import debounce from 'lodash/debounce';
+import throttle from 'lodash/throttle';
+import StyledModal from 'src/components/Modal';
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:26]: I think you can just name it as `Modal` in this case

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+  tables: any[];
+  editorQueries: QueryEditor[];
+  dataPreviewQueries: object[];
+  queryEditorId: string;
+  hideLeftBar?: boolean;
+  defaultQueryLimit: number;
+  maxRow: number;
+  displayLimit: number;
+  saveQueryWarning?: string;
+  scheduleQueryWarning?: string;
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:113]: Do you think that for `actions` `tables` `dataPreviewQueries` we can be a little more specific for their types?

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+}
+
+const LimitSelectStyled = styled.span`
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:116]: Nit but for consistency I think we should name it `StyledLimitSelect`

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+  const [autocompleteEnabled, setAutocompleteEnabled] = useState(
+    getItem(LocalStorageKeys.sqllab__is_autocomplete_enabled, true),
+  );
+  const [showCreateAsModal, setShowCreateAsModal] = useState(false);
+  const [createAs, setCreateAs] = useState('');
+
+  const sqlEditorRef = useRef(null);
+  const northPaneRef = useRef(null);
+
+  useState(() => {
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:201]: Are you sure that you wanted to use `useState` here?

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+
+  const renderDropdown = () => {
+    const qe = queryEditor;
+    const successful = latestQuery?.state === 'success';
+    const scheduleToolTip = successful
+      ? t('Schedule the query periodically')
+      : t('You must run the query successfully first');
+    return (
+      <Menu onClick={this.handleMenuClick} style={{ width: 176 }}>
+        <Menu.Item style={{ display: 'flex', justifyContent: 'space-between' }}>
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:517]: We should use emotion for styling both `Menu` and `Menu.Item` and avoid inline styles wherever possible

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+    );
+  };
+
+  const createViewModalTitle =
+    createAs === CtasEnum.VIEW ? 'CREATE VIEW AS' : 'CREATE TABLE AS';
+
+  const createModalPlaceHolder =
+    createAs === CtasEnum.VIEW
+      ? 'Specify name to CREATE VIEW AS schema in: public'
+      : 'Specify name to CREATE TABLE AS schema in: public';
```
Comment (geido) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:689]:
```suggestion
      ? t('Specify name to CREATE VIEW AS schema in: public')
      : t('Specify name to CREATE TABLE AS schema in: public');
```

**Review by geido — COMMENTED** [review]
Hey @EugeneTorap thanks for the work that you put into this! Please find below some initial comments from a first look.

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx (hunk 1 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx]: image
@geido Dropdown UI is changed

Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx]: I think I should revert these changes

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx (hunk 3 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:109]: It's better do in next PR

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx (hunk 5 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:197]: This `useState` for `UNSAFE_componentWillMount` replacement

File: superset-frontend/src/SqlLab/components/SqlEditor/index.tsx (hunk 3 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:109]: @geido Do we have types for all sqlLab `actions` & `dataPreviewQueries`?

Comment (geido) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.tsx:113]: @EugeneTorap if you can't find any reference in the code, we probably don't have but I am happy to do that in a related PR!

**Comment by geido** [comment]
I am now running the CI. WE should spin up a test env as soon as it passes

**Review by lyndsiWilliams — CHANGES_REQUESTED** [review]
Please fix the [broken frontend tests](https://github.com/apache/superset/runs/5165997034?check_suite_focus=true).

**Pull request closed without merging**